src/processor/operator/generic_extend.cpp

```diff
@@ -0,0 +1,127 @@
+#include "processor/operator/generic_extend.h"
+
+namespace kuzu {
+namespace processor {
+
+shared_ptr<ResultSet> GenericExtend::init(ExecutionContext* context) {
+    resultSet = PhysicalOperator::init(context);
+    inVector = resultSet->getValueVector(inVectorPos);
+    auto outDataChunk = resultSet->dataChunks[outVectorPos.dataChunkPos];
+    outVector = make_shared<ValueVector>(NODE_ID, context->memoryManager);
+    outDataChunk->insert(outVectorPos.valueVectorPos, outVector);
+    resultSet->initListSyncState(outVectorPos.dataChunkPos);
+    for (auto& list : lists) {
+        listHandles.push_back(
+            make_shared<ListHandle>(*resultSet->getListSyncState(outVectorPos.dataChunkPos)));
+    }
+    nextListIdx = lists.size();
+    nextColumnIdx = columns.size();
+    currentList = nullptr;
+    return resultSet;
+}
+
+bool GenericExtend::getNextTuplesInternal() {
```

Comment:
One thing I would try is as follows:
a) for implementation of `findMoreOutput()`, merge `findInCurrentList` and `findInRemainingLists` into one function `scanFromList`, which will handle both cases.
b) move the block
```
    if (findMoreOutput()) {
         return true;
     }
```
into the while loop. So each time you check if you can read with current states (currentNodeOffset, currentList, nextList/ColumnIdx), if not, call getNextTuple, then reset states, and go back to the beginning to read with current states again.
3. thus, you don't need findOutput() inside the while condition.
I think, if possible, this will simplify the whole control logic.

Reply:
Done